publication-commons/src/main/java/no/unit/nva/publication/service/impl/ApprovalAssignmentServiceForImportCandidate.java

```diff
@@ -0,0 +1,118 @@
+                                                         ImportCandidate importCandidate) {
+        return getCustomerForCorrespondenceContributor(customers, importCandidate)
+                   .or(() -> getCustomerForContributorWithLowestSequenceNumber(importCandidate, customers));
+    }
+
+    private Optional<CustomerDto> getCustomerForCorrespondenceContributor(Collection<CustomerDto> customers, ImportCandidate importCandidate) {
+        return importCandidate.getEntityDescription()
+                   .getContributors().stream()
+                   .filter(Contributor::isCorrespondingAuthor)
+                   .sorted(Comparator.comparing(Contributor::getSequence,
+                                                Comparator.nullsLast(Comparator.naturalOrder())))
+                   .map(contributor -> getCustomerForContributor(contributor, customers))
+                   .filter(Optional::isPresent)
+                   .map(Optional::get)
+                   .findFirst();
+    }
+
+    private Optional<CustomerDto> getCustomerForContributor(Contributor contributor, Collection<CustomerDto> customers) {
+        return contributor.getAffiliations().stream()
+                   .filter(Organization.class::isInstance)
+                   .map(Organization.class::cast)
+                   .map(Organization::getId)
+                   .filter(Objects::nonNull)
+                   .flatMap(cristinId -> getCustomerByCristinId(customers, cristinId))
+                   .findFirst();
+    }
+
+    private Optional<CustomerDto> getCustomerForContributorWithLowestSequenceNumber(ImportCandidate importCandidate,
+                                                                                    Collection<CustomerDto> customers) {
+        return importCandidate.getEntityDescription()
+                   .getContributors().stream()
+                   .sorted(Comparator.comparing(Contributor::getSequence,
+                                                Comparator.nullsLast(Comparator.naturalOrder())))
+                   .map(contributor -> getCustomerForContributor(contributor, customers))
+                   .filter(Optional::isPresent)
+                   .map(Optional::get)
+                   .findFirst();
+    }
+
+    private Stream<CustomerDto> getCustomerByCristinId(Collection<CustomerDto> customers, URI cristinId) {
+        return customers.stream()
+                   .filter(customerDto -> cristinIdMatchesCustomer(cristinId, customerDto))
+                   .findFirst().stream();
+    }
+
+    private boolean cristinIdMatchesCustomer(URI organizationId, CustomerDto customerDto) {
+        return extractOrganizationNumber(organizationId).equals(extractOrganizationNumber(customerDto.cristinId()));
+    }
+
+    private static String extractOrganizationNumber(URI organizationId) {
+        return UriWrapper.fromUri(organizationId).getLastPathElement().split(CRISTIN_ID_SEPARATOR)[0];
+    }
```

Comment:
Maybe I should consider util @torbjokv implemented when generating `curratingInstitutions` for publication?
Is s3 file with all organizations always updated?

Reply:
I would certainly advise against this approach :D